feat: EPI-715: PatientData question type pulling through id instead of name

khoi-bes opened the pull request:
### Changes

- Breaking changes:
 + New components for PatientData questions in view-only mode. This component will be used for PatientData (which is able to write to patient) in the form response.
 + Fix the Patient source autocomplete question (.i.e mother, father,..) issue
 + Update translations for PatientData questions result
 + Refactor logic for SurveyAnswer question (which has source = PatientData question)

### Deploys

- [x] **Deploy to Tamanu Internal** <!-- #deploy -->

### Tests

- [ ] **Run E2E Tests** <!-- #e2e -->

### Remember to...

- ...write or update tests
- ...add UI screenshots and **testing notes** to the Linear issue
- ...add any **manual upgrade steps** to the Linear issue
- ...update the [config reference](https://beyond-essential.slab.com/posts/reference-config-file-0c70ukly), [settings reference](https://beyond-essential.slab.com/posts/reference-settings-0blw1x2q), or any [relevant runbook(s)](https://beyond-essential.slab.com/topics/runbooks-bs04ml6c)
- ...call out additions or changes to **config files** for the deployment team to take note of

<!-- Thank you! -->

khoi-bes committed:
feat: EPI-715: patientdata-question-type-pulling-through-id-instead-of-name

khoi-bes committed:
fix(SurveyForm): Safeguard access to customPatientFieldValues with optional chaining

khoi-bes committed:
feat(surveyResponse): Add motherId and fatherId columns to patient_additional_data migration and improve error messaging in surveyResponse route

jaskfla commented:
bugbot run

passcod commented on packages/constants/src/surveys.ts:
```diff
@@ -121,18 +132,24 @@ export const VITAL_CHARTS = {
 
 type PatientDataFieldLocationsType = {
-  [key: string]: Array<string>;
+  [key: string]: [string, string, Record<string, string>] | [string, string];
```
preference would be to change this to some kind of descriptive interface, but probably out of scope; at least can you add a comment describing what the different string fields of the array do please

passcod commented on packages/shared/src/reports/utilities/transformAnswers.js:
```diff
@@ -1,22 +1,33 @@
 
-// also update getNameColumnForModel in /packages/mobile/App/ui/helpers/fields.ts when this changes
-function getNameColumnForModel(modelName) {
+// also update getDisplayNameForModel in /packages/mobile/App/ui/helpers/fields.ts when this changes
```
similarly as constants, we can used shared package in mobile directly

passcod reviewed (changes requested):
a pretty large PR; could maybe do with some breaking apart. regardless, needs package reuse from shared/constants instead of doubling up. otherwise okay

khoi-bes committed:
Merge branch 'main' into feat/epi-715-patientdata-question-type-pulling-through-id-instead-of-name

khoi-bes committed:
chore: update CI workflow and clean up constants

- Added a build step to the CI workflow for shared components.
- Removed unused patientFields and surveys constants from the mobile app.
- Updated imports in various files to reference the new constants structure.

khoi-bes committed:
Merge branch 'main' into feat/epi-715-patientdata-question-type-pulling-through-id-instead-of-name

passcod commented on packages/facility-server/app/routes/apiv1/surveyResponse.js:
```diff
@@ -141,3 +145,45 @@ surveyResponse.put(
+      return;
+    }
+
+    const [modelName, fieldName] = PATIENT_DATA_FIELD_LOCATIONS[column] || [];
```
we should probably return an error if the column is not available to be queried (ie not in the constant) rather than returning null

passcod commented on packages/facility-server/app/routes/apiv1/surveyResponse.js:
```diff
@@ -141,3 +145,45 @@ surveyResponse.put(
+
+    if (!modelName) {
+      res.json({
+        data: value,
```
I'm also unsure why this is a thing. Why do we return the passed value if we don't find a model?

passcod commented on packages/mobile/App/migrations/1756663661549-addMotherFatherIdToPatientAdditionalData.ts:
```diff
@@ -0,0 +1,62 @@
+      );
+    }
+
+        // Note: Foreign key constraints are handled by TypeORM at the application level
+    // Not creating database-level foreign keys to avoid constraint violations during data sync
```
```suggestion
    // Note: Foreign key constraints are handled by TypeORM at the application level
    // Not creating database-level foreign keys to avoid constraint violations during data sync
```

passcod reviewed (approved):
a few more things but happy to call those optional to passing

khoi-bes committed:
refactor: streamline error handling and code formatting in surveyResponse and survey utility files

- Improved error handling in surveyResponse.js by throwing an error for invalid columns.
- Refactored arrow function syntax for consistency in survey.jsx, enhancing readability.
- Cleaned up redundant code and ensured uniformity in function definitions across the file.

khoi-bes committed:
Merge branch 'main' into feat/epi-715-patientdata-question-type-pulling-through-id-instead-of-name

khoi-bes committed:
Merge branch 'main' into feat/epi-715-patientdata-question-type-pulling-through-id-instead-of-name

khoi-bes committed:
refactor: simplify getEnumTranslation function by removing unused enumFallback parameter
